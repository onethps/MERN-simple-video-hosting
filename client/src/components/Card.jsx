import React, {useEffect, useState} from 'react';
import styled from 'styled-components';
import {Link} from 'react-router-dom';
import {format} from 'timeago.js';
import {instance} from "api/config";

const Container = styled.div`
  width: ${(props) => (props.type === 'sm' ? '100%' : '350px')};
  margin-bottom: ${(props) => (props.type === 'sm' ? '10px' : '45px')};
  cursor: pointer;
  display: ${(props) => props.type === 'sm' && 'flex'};
  gap: 20px;
`;

const ThumbNail = styled.img`
  min-width: ${(props) => (props.type === 'sm' ? '50%' : '100%')};
  object-fit: cover;
  height: ${(props) => (props.type === 'sm' ? '120px' : '196px')};
  border-radius: 20px;
  background: grey;
`;

const Description = styled.div``;

const Title = styled.h1`
  font-size: 16px;
  color: ${({theme}) => theme.text};
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const Author = styled.h2`
  font-size: 14px;
  font-weight: 400;
  color: ${({theme}) => theme.text};
`;

const Views = styled.div`
  font-size: 12px;
  color: ${({theme}) => theme.text};
`;

const LoadingBlock = styled.div`
  margin-top: 60px;
`;

function Card({type, video}) {
  const [channel, setChannel] = useState(null);

  useEffect(() => {
    const fetchChannel = async () => {
      const {data} = await instance.get(`/users/find/${video.userId}`);
      setChannel(data);
    };

    fetchChannel();
  }, [video.userId]);

  return (
    <Link to={`/video/${video._id}`} style={{textDecoration: 'none'}}>
      <Container type={type}>
        <ThumbNail type={type} src={video.imgUrl}/>
        <Description>
          <Title>{video?.title}</Title>
          <Author>{channel?.name}</Author>
          <Views>
            {video?.views} views - {format(video.createdAt)}
          </Views>
        </Description>
      </Container>
    </Link>
  );
}


export {Card};
