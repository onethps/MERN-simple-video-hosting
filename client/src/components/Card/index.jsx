import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { instance } from 'api/config';
import { format } from 'timeago.js';
import {
  Container,
  ThumbnailBox,
  Desc,
  Thumbnail,
  Title,
  Views,
  Author,
} from './styles/card';

const Card = ({ type, video }) => {
  const [channel, setChannel] = useState(null);

  useEffect(() => {
    const fetchChannel = async () => {
      const { data } = await instance.get(`/users/find/${video.userId}`);
      setChannel(data);
    };

    fetchChannel();
  }, [video.userId]);

  return (
    <Container type={type}>
      <Link type={type} to={`/video/${video._id}`} style={{ textDecoration: 'none' }}>
        <ThumbnailBox type={type}>
          <Thumbnail type={type} src={video.imgUrl} />
        </ThumbnailBox>
        <Desc>
          <Title>{video?.title}</Title>
          <Author>{channel?.name}</Author>
          <Views>
            {video?.views} views - {format(video?.createdAt)}
          </Views>
        </Desc>
      </Link>
    </Container>
  );
};

export default Card;
