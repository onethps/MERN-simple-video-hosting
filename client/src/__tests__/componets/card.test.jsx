import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { format } from 'timeago.js';
import Card from '../../components/Card/index';

describe('<Card />', () => {
  const type = 'sm';

  const user = {
    _id: '55',
    name: 'Victor',
  };

  const cardVideos = [
    {
      _id: '1',
      imgUrl: 'https://www.google.com/',
      title: 'Video 1',
      views: '10',
      userId: '55',
      createdAt: new Date().toLocaleString(),
    },
    {
      _id: '2',
      imgUrl: 'https://www.google.com/',
      title: 'Video 2',
      views: '50',
      userId: '12',
      createdAt: new Date().toLocaleString(),
    },
  ];

  it('renders the <Card /> with populated data', () => {
    const { getByText, container } = render(
      <Router>
        <>
          {cardVideos.map((video) => {
            return (
              <Card key={video._id} type={type}>
                <Card.Link to={`/video/${video?._id}`}>
                  <Card.ThumbnailBox>
                    <Card.Thumbnail src={video?.imgUrl} />
                  </Card.ThumbnailBox>
                  <Card.Base>
                    <Card.Title>{video?.title}</Card.Title>
                    <Card.Author>{user?.name}</Card.Author>
                    <Card.ViewsAndDate>
                      {video?.views} views - {format(video?.createdAt)}
                    </Card.ViewsAndDate>
                  </Card.Base>
                </Card.Link>
              </Card>
            );
          })}
        </>
      </Router>,
    );

    expect(getByText('Video 1')).toBeInTheDocument();
    expect(getByText('50 views - just now')).toBeInTheDocument();
    expect(getByText('Video 2')).toBeTruthy();
    expect(getByText('10 views - just now')).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
