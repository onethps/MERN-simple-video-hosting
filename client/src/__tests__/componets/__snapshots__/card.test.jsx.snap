// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders the <Card /> with populated data 1`] = `
<div
  class="sc-bczRLJ bbLVXS"
  type="sm"
>
  <a
    class="sc-gsnTZi dcGqJy"
    href="/video/1"
  >
    <div
      class="sc-dkzDqf ezTOpN"
    >
      <img
        class="sc-hKMtZM ymCzX"
        src="https://www.google.com/"
      />
    </div>
    <div
      class="sc-jSMfEi"
    >
      <h1
        class="sc-eCYdqJ kkkzZv"
      >
        Video 1
      </h1>
      <h2
        class="sc-gKXOVf gTEanZ"
      >
        Victor
      </h2>
      <div
        class="sc-iBkjds beyYmH"
      >
        10
         views - 
        just now
      </div>
    </div>
  </a>
</div>
`;
